Add VersionChecker to detect and warn about newer Bicep CLI installations

polatengin opened the pull request:
Fixes #5070

## Summary

This PR updates `bicep --version` to provide richer version diagnostics for users who may have multiple Bicep CLI installations or may be running an older CLI than the latest released version.

The command now reports:

- current Bicep CLI version
- git commit SHA, when available
- OS, OS version, and architecture
- latest Bicep release from GitHub, when reachable
- newer Bicep CLI installations found on the local machine

## Details

- Adds a `VersionChecker` helper that scans common Bicep install locations and `PATH` entries for `bicep` / `bicep.exe`.
- Uses `FileVersionInfo` metadata to inspect installed Bicep binaries instead of launching child `bicep` processes.
- Deduplicates normalized paths and ignores inaccessible directories, invalid binaries, and unparsable versions.
- Adds `GitHubLatestReleaseClient` to query the latest Bicep release from GitHub and compare it with the current CLI version.
- Falls back to `Latest GitHub release: unavailable` if the GitHub request fails or returns an invalid release payload.
- Extends `IEnvironment` with OS version and operating system architecture so `--version` can include useful troubleshooting context.
- Adds unit and integration coverage for version parsing, install-location discovery, newer-install detection, GitHub release parsing, and the updated `--version` output.

Well-known installation locations checked:

`~/.bicep/bin` (default location)
`~/.azure/bin` (Azure CLI installation location)

Windows-specific:
- `C:\Program Files\Bicep CLI`
- `C:\Program Files (x86)\Bicep CLI`
- _All directories in the PATH environment variable_

Linux-specific:
- `/usr/local/bin`
- `/usr/bin/`
- _All directories in the PATH environment variable_

## Example Output

```text
Bicep CLI version: 0.30.23
Git commit SHA: abcdef
OS: Linux
OS version: Test OS 1.2.3
Architecture: X64

Latest GitHub release: 0.40.0
A newer Bicep CLI release is available: https://github.com/Azure/bicep/releases/tag/v0.40.0

Newer Bicep CLI installation(s) found:
  - Version 0.40.0 (Git commit SHA: cdefab) at /usr/local/bin/bicep
```

To check the output of the case there are other versions exist in the system;

```shell
tmpdir=$(mktemp -d) && dotnet new console -n bicep -o "$tmpdir/fake-src" >/tmp/bicep-version-test-new-dll.log && dotnet build "$tmpdir/fake-src/bicep.csproj" -c Release -p:Version=1.1.1 -p:FileVersion=1.1.1 -p:AssemblyVersion=1.1.1 -p:InformationalVersion=1.1.1+abcdef >/tmp/bicep-version-test-build-dll.log && mkdir -p "$tmpdir/path-bin" && cp "$tmpdir/fake-src/bin/Release/net10.0/bicep.dll" "$tmpdir/path-bin/bicep" && PATH="$tmpdir/path-bin:$PATH" dotnet run --project ./src/Bicep.Cli/Bicep.Cli.csproj -- --version
```

Output;

```shell
Bicep CLI version: 0.43.11
Git commit SHA: 3ed2670d9d
OS: Linux
OS version: Debian GNU/Linux 11 (bullseye)
Architecture: X64

Latest GitHub release: 0.43.8
You are running the latest Bicep CLI release.

Newer Bicep CLI installation(s) found:
  - Version 1.1.1 (Git commit SHA: abcdef) at /tmp/tmp.dv5MtanF2E/path-bin/bicep
```

## Checklist

- [x] I have read and adhere to the [contribution guide](https://github.com/Azure/bicep/blob/main/CONTRIBUTING.md).
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/Azure/bicep/pull/18375)

polatengin committed:
Add VersionChecker to detect and warn about newer Bicep CLI installations

VersionChecker helper scans well-known install locations, runs `bicep --version` to parse versions, and reports any installations newer than the running CLI.

levimatheri commented on src/Bicep.Cli/Helpers/VersionChecker.cs:
```diff
@@ -0,0 +1,228 @@
+        catch
+        {
+            return null;
+        }
```
Could we consider using [`FileVersionInfo`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.fileversioninfo?view=net-9.0) here, rather than launching a process and parsing the output? i.e.

```csharp
var fileVersionInfo = System.Diagnostics.FileVersionInfo.GetVersionInfo(bicepPath);
var version = fileVersionInfo.FileVersion;
```

polatengin replied:
agreed and done, thanks 👍

anthony-c-martin reviewed (changes requested):
This is a pretty major behavioral change, and could have serious performance implications that we need to think through, as well as product design implications. At minimum, I think we need to cover this as a topic in Bicep Discussions before continuing on with the change.

Generally, I think my feeling with doing this work is that the cons outweigh the pros. Issue #5070 was created a long time ago, and since then, I don't think it's been as much of a problem as we imagined it being.

polatengin committed:
Use FileVersionInfo to get Bicep executable version instead of launching process

Avoid spawning a bicep process and parsing its stdout. Check file existence earlier and read FileVersionInfo.FileVersion, parsing it with Version.TryParse and returning null for missing/invalid versions.

polatengin committed:
Run version check asynchronously and limit scan scope

Make VersionChecker.CheckForNewerVersions non-blocking by introducing
CheckForNewerVersionsAsync that runs the scan in a background task with a
100ms timeout and cancellation token. Add a shouldCheck parameter so
callers can skip the scan for lightweight commands or when output is
redirected. Make scanning cancellation-aware and restrict file searches
to the top directory only. Remove scanning of PATH entries from
well-known locations.

Update Program to compute whether the check should run (ShouldCheckForNewerVersions)
and invoke the new async method.

polatengin committed:
Add VersionChecker tests and make internals testable

- Add VersionCheckerTests with coverage for:
  - asynchronous check behavior (shouldCheck flag)
  - no/newer/older/multiple versions detection
  - platform-specific install locations (Linux/Windows)
  - non-existent directories, invalid versions, and cancellation handling
- Add TestableVersionChecker to mock Bicep file versions and normalize paths for reliable tests
- Make FindNewerVersions and GetWellKnownInstallLocations public, and make GetBicepVersion protected virtual to allow overriding in tests

polatengin committed:
Normalize path handling in VersionChecker tests for cross-platform compatibility

polatengin committed:
Enhance Version Checking and CLI Version Output

- Added a new method to VersionChecker to find newer Bicep CLI installations and return their versions along with Git commit SHAs.
- Updated the RootCommand to print the current Bicep CLI version, OS details, and any newer installations found.
- Refactored version parsing logic to handle Git commit SHAs and normalize version formats.
- Improved test coverage for version checking, including scenarios for newer versions and OS-specific paths.
- Enhanced the IEnvironment interface to expose operating system version and architecture information.

polatengin committed:
Merge remote-tracking branch 'origin/main' into polatengin/5070-try-to-identify-when-two-versions-of-bicep-are-installed-on-the-system

polatengin committed:
Update version output format in PrintVersion method

polatengin committed:
Update version output format in Bicep CLI integration tests

polatengin committed:
Implement GitHub latest release version checking in Bicep CLI

polatengin committed:
Update expected path in FindNewerVersions test for Bicep installation
